File: launch/interface/src/main/java/xsbti/ApplicationID.java
package xsbti;

import java.io.File;

/**
 * This represents an identification for the sbt launcher to load and run
 * an sbt launched application using ivy.
 */
public interface ApplicationID
{
	/**
	 * @return
	 *   The Ivy orgnaization / Maven groupId where we can find the application to launch.
	 */
	public String groupID();
	/**
	 * @return
	 *    The ivy module name / Maven artifactId where we can find the application to launch.
	 */
	public String name();
	/**
	 * @return
	 *    The ivy/maven version of the module we should resolve.
	 */
	public String version();

	/**
	 * @return
	 *    The fully qualified name of the class that extends xsbti.AppMain
	 */
	public String mainClass();
	/**
	 * @return
	 *    Additional ivy components we should resolve with the main application artifacts.
	 */
	public String[] mainComponents();
	/**
	 * @deprecated
	 * This method is no longer used if the crossVersionedValue method is available.
	 * 
	 * @return
	 *    True if the application is cross-versioned by binary-compatible version string,
	 *    False if there is no cross-versioning.
	 */
	@Deprecated
	public boolean crossVersioned();
	
	/**
	 * 
	 * @since 0.13.0
	 * @return
	 *    The type of cross-versioning the launcher should use to resolve this artifact.
	 */
	public CrossValue crossVersionedValue();
	
	/** Files to add to the application classpath. */
	public File[] classpathExtra();
}

File: sbt-launcher/src/main/scala/coursier/sbtlauncher/ApplicationID.scala
package coursier.sbtlauncher

import java.io.File

final case class ApplicationID(
  groupID: String,
  name: String,
  version: String,
  mainClass: String,
  mainComponents: Array[String],
  crossVersioned: Boolean,
  crossVersionedValue: xsbti.CrossValue,
  classpathExtra: Array[File]
) extends xsbti.ApplicationID {

  assert(crossVersioned == (crossVersionedValue != xsbti.CrossValue.Disabled))

  def disableCrossVersion(scalaVersion: String): ApplicationID =
    crossVersionedValue match {
      case xsbti.CrossValue.Disabled =>
        this
      case xsbti.CrossValue.Binary =>
        val scalaBinaryVersion = scalaVersion.split('.').take(2).mkString(".")
        copy(
          crossVersioned = false,
          crossVersionedValue = xsbti.CrossValue.Disabled,
          version = s"${version}_$scalaBinaryVersion"
        )
      case xsbti.CrossValue.Full =>
        copy(
          crossVersioned = false,
          crossVersionedValue = xsbti.CrossValue.Disabled,
          version = s"${version}_$scalaVersion"
        )
    }
}

object ApplicationID {
  def apply(id: xsbti.ApplicationID): ApplicationID =
    id match {
      case id0: ApplicationID => id0
      case _ =>
        ApplicationID(
          id.groupID(),
          id.name(),
          id.version(),
          id.mainClass(),
          id.mainComponents(),
          id.crossVersionedValue() != xsbti.CrossValue.Disabled,
          id.crossVersionedValue(),
          id.classpathExtra()
        )
    }
}
